File: MinimumInASortedRotatedArray.java
// Binary search; if the element at mid is greater than the element at the end, the min value lies in the
// second half, update start index, else min element lies in the first half; at the end, return the low/start index
// Time complexity: O(log n)
// Space complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No




public class MinimumInASortedRotatedArray {
	
	    public int findMin(int[] nums) {
	        int mid = 0;
	        int len = nums.length-1;
	        int low = 0;
	        int high = len;
	        while(low < high){
	            mid = (low+high)/2;
	            if(nums[mid] > nums[high]){
	                low = mid+1;
	            }
	            else
	                high = mid;
	        }
	        return nums[low];
	    
	}

}

File: Minimum_in_a_sorted_rotated_array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
       int n = nums.size()-1, low = 0, high = nums.size()-1;
       if(n ==-1) return -1;
       if(n==0) return nums[0];
        while(low<=high){
            int mid = low + (high-low)/2;
        if(nums[low] <= nums[high])return nums[low];
        if((mid == 0 || nums[mid] < nums[mid-1]) && (nums[mid] < nums[mid+1])){
                return nums[mid];
         }
         else if(nums[mid]< nums[low]) {
             high = mid -1;
         }
         else low = mid+1;

        }
   return -1;
    } 
};
